src/pages/ExamsPage/components/AddAllowanceModal.jsx

```diff
@@ -0,0 +1,225 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import {
+  ModalDialog,
+  Form,
+  ActionRow,
+  StatefulButton,
+} from '@openedx/paragon';
+import { useIntl } from '@edx/frontend-platform/i18n';
+
+import * as constants from '../../../data/constants';
+import { useExamsData, useCreateAllowance, useButtonStateFromRequestStatus } from '../hooks';
+import messages from '../messages';
+
+const AddAllowanceModal = ({ isOpen, close }) => {
+  const { examsList } = useExamsData();
+  const [displayExams, setDisplayExams] = useState([]);
+  const [form, setForm] = useState({ 'allowance-type': 'additional-minutes' });
+  const [learnerFieldError, setLearnerFieldError] = useState(false);
+  const [examFieldError, setExamFieldError] = useState(false);
+  const [additionalTimeError, setAdditionalTimeError] = useState(false);
+  const createAllowanceRequestStatus = useButtonStateFromRequestStatus(constants.RequestKeys.createAllowance);
+  const createAllowance = useCreateAllowance();
+  const { formatMessage } = useIntl();
+
+  const addAllowanceButtonProps = {
+    labels: {
+      default: formatMessage(messages.addAllowanceButtonDefaultLabel),
+      pending: formatMessage(messages.addAllowanceButtonPendingLabel),
+      complete: formatMessage(messages.addAllowanceButtonCompleteLabel),
+      error: formatMessage(messages.addAllowanceButtonErrorLabel),
+    },
+    variant: 'primary',
+  };
+
+  const handleExamTypeChange = () => {
+    if (form['exam-type'] === 'proctored') {
+      setDisplayExams(proctoredExams);
+    } else if (form['exam-type'] === 'timed') {
+      setDisplayExams(timedExams);
+    }
+  };
+
+  const proctoredExams = [];
+  const timedExams = [];
+
+  useEffect(() => {
+    for (let i = 0; i < examsList.length; i++) {
+      const exam = examsList[i];
+      if (exam.examType === 'proctored') {
+        proctoredExams.push(exam);
+      } else {
+        timedExams.push(exam);
+      }
+      const defaultExamType = proctoredExams.length > 0 ? 'proctored' : 'timed';
+      setForm(prev => ({ ...prev, 'exam-type': defaultExamType }));
+      handleExamTypeChange();
+    }
+  }, [examsList]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
Nothing stands out to me about the hook that would cause it not to run. Have you confirmed that it's not running, or is it running and not doing anything? Maybe there's an issue with your set up (e.g. `examsList` is empty)?

However, my initial thought is that this doesn't need to be in a `useEffect` hook, based on [You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect). When you say "it causes too many renders" when you move it out of the `useEffect` hook, what do you mean by that?

If necessary, you could use the `useMemo` hook to memoize the results of the computation to avoid recomputing it unless `examsList` changes, but you'd have to refactor your hook a little to return something that can be memoized (e.g. the `proctoredExams` and `timedExams` variables).

What do you think?

Reply:
If I run tests when this code is outside of a hook, the tests fail because of too many re-renders. The line that's specifically causing too many rerenders is `setForm(prev => ({ ...prev, 'exam-type': defaultExamType }));`. If I comment that line out, the tests pass.

Reply:
I'll pull down your code and take a look. I'll follow up soon.